sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java

```diff
@@ -587,61 +742,117 @@ private long recomputeIsRealtime(ImmutableSet<DruidServerMetadata> servers)
     return retVal;
   }
 
+  /**
+   * This method is not thread-safe and must be used only in unit tests.
+   */
   @VisibleForTesting
   void setAvailableSegmentMetadata(final SegmentId segmentId, final AvailableSegmentMetadata availableSegmentMetadata)
   {
-    synchronized (lock) {
-      TreeMap<SegmentId, AvailableSegmentMetadata> dataSourceSegments = segmentMetadataInfo.computeIfAbsent(
-          segmentId.getDataSource(),
-          x -> new TreeMap<>(SEGMENT_ORDER)
-      );
-      if (dataSourceSegments.put(segmentId, availableSegmentMetadata) == null) {
-        totalSegments++;
-      }
+    final ConcurrentSkipListMap<SegmentId, AvailableSegmentMetadata> dataSourceSegments = segmentMetadataInfo
+        .computeIfAbsent(
+            segmentId.getDataSource(),
+            k -> new ConcurrentSkipListMap<>(SEGMENT_ORDER)
+        );
+    if (dataSourceSegments.put(segmentId, availableSegmentMetadata) == null) {
+      totalSegments++;
     }
   }
 
-  protected DruidTable buildDruidTable(final String dataSource)
+  /**
+   * This is a helper method for unit tests to emulate heavy work done with {@link #lock}.
+   * It must be used only in unit tests.
```

Comment:
nit: maybe nice to move all of these 'only for testing' methods to the end of this file or something to get them out of the way.